### PR TITLE
Implement parallel block eviction

### PR DESCRIPTION
**Pull request by ggezer**

### CODE REVIEW

**Comment by ggezer** [comment]
Jenkins, test this please

File: core/common/src/main/java/alluxio/conf/PropertyKey.java [hunk 1]
```diff
@@ -2294,6 +2294,14 @@ public String toString() {
           .setConsistencyCheckLevel(ConsistencyCheckLevel.WARN)
           .setScope(Scope.WORKER)
           .build();
+  public static final PropertyKey WORKER_NETWORK_BLOCK_EVICTION_THREADS =
+      new Builder(Name.WORKER_NETWORK_BLOCK_EVICTION_THREADS)
+          .setDefaultSupplier(() -> Runtime.getRuntime().availableProcessors(),
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/conf/PropertyKey.java]: Is this a reasonable default?

File: core/common/src/main/java/alluxio/conf/PropertyKey.java [hunk 2]
```diff
@@ -4352,6 +4360,8 @@ private static String javadocLink(String fullyQualifiedClassname) {
         "alluxio.worker.network.block.reader.threads.max";
     public static final String WORKER_NETWORK_BLOCK_WRITER_THREADS_MAX =
         "alluxio.worker.network.block.writer.threads.max";
+    public static final String WORKER_NETWORK_BLOCK_EVICTION_THREADS =
+        "alluxio.worker.network.block.eviction.threads";
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/conf/PropertyKey.java]: Does this fit the `network` category?

File: core/server/worker/src/main/java/alluxio/worker/block/TieredBlockStore.java [hunk 3]
```diff
@@ -691,39 +703,62 @@ private void freeSpaceInternal(long sessionId, long availableBytes, BlockStoreLo
...
+                for (BlockStoreEventListener listener : mBlockStoreEventListeners) {
+                  listener.onMoveBlockByWorker(sessionId, blockId, moveResult.getSrcLocation(),
+                      newLocation);
+                }
+              }
+              return true;
             }
+          } catch (InvalidWorkerStateException e) {
+            // Evictor is not working properly
+            LOG.error("Failed to demote blockId {}, this is temp block", blockId);
```
Comment (calvinjia) [inline core/server/worker/src/main/java/alluxio/worker/block/TieredBlockStore.java]: Should this be warn or should we do something to handle this error?

File: core/server/worker/src/main/java/alluxio/worker/block/TieredBlockStore.java [hunk 4]
```diff
@@ -691,39 +703,62 @@ private void freeSpaceInternal(long sessionId, long availableBytes, BlockStoreLo
...
+          } catch (BlockAlreadyExistsException e) {
+            LOG.info("Failed to demote blockId {}, it could be already moved", blockId);
+          } catch (BlockDoesNotExistException e) {
+            LOG.info("Failed to demote blockId {}, it could be already deleted", blockId);
           }
-        }
+          return false;
+        });
```
Comment (calvinjia) [inline core/server/worker/src/main/java/alluxio/worker/block/TieredBlockStore.java]: This block of logic is pretty involved, perhaps make a separate method?

File: core/server/worker/src/main/java/alluxio/worker/block/TieredBlockStore.java [hunk 5]
```diff
@@ -691,39 +703,62 @@ private void freeSpaceInternal(long sessionId, long availableBytes, BlockStoreLo
...
+      }
+      // Execute move orders in thread pool.
+      try {
+        List<Future<Boolean>> moveResults = mEvictionExecutor.invokeAll(moveTasks);
+        long failCount = moveTasks.size() - moveResults.stream().filter(
+            (fut) -> {
+              try {
+                return fut.get();
+              } catch (Exception e) {
+                return false;
```
Comment (calvinjia) [inline core/server/worker/src/main/java/alluxio/worker/block/TieredBlockStore.java]: Do we need to distinguish between the future failures (which return false) and the future exceptions (which are caught here and will return false)?

File: core/server/worker/src/main/java/alluxio/worker/block/TieredBlockStore.java [hunk 6]
```diff
@@ -691,39 +703,62 @@ private void freeSpaceInternal(long sessionId, long availableBytes, BlockStoreLo
...
+              }
+            }).count();
+        LOG.info("Finished moving:{} blocks to tier: {}. Failure count: {} ", moveTasks.size(),
```
Comment (calvinjia) [inline core/server/worker/src/main/java/alluxio/worker/block/TieredBlockStore.java]: space between `:` and `{`?

File: core/server/worker/src/main/java/alluxio/worker/block/TieredBlockStore.java [hunk 7]
```diff
@@ -691,39 +703,62 @@ private void freeSpaceInternal(long sessionId, long availableBytes, BlockStoreLo
...
+            mStorageTierAssoc.getAlias(tierOrdinal), failCount);
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        throw new RuntimeException("Interrupted while executing eviction move orders.");
```
Comment (calvinjia) [inline core/server/worker/src/main/java/alluxio/worker/block/TieredBlockStore.java]: Should we attempt to cancel or interrupt the tasks running in the eviction executor? Also, is throwing runtime exception here handled properly by the callers?

File: core/common/src/main/java/alluxio/conf/PropertyKey.java (hunk 1 above)
Comment (ggezer) [reply core/common/src/main/java/alluxio/conf/PropertyKey.java]: I found this to yield min eviction duration.

File: core/server/worker/src/main/java/alluxio/worker/block/TieredBlockStore.java (hunk 3 above)
Comment (ggezer) [reply core/server/worker/src/main/java/alluxio/worker/block/TieredBlockStore.java]: This is not expected so I'd leave it ERROR based on above comment.

File: core/server/worker/src/main/java/alluxio/worker/block/TieredBlockStore.java (hunk 7 above)
Comment (ggezer) [reply core/server/worker/src/main/java/alluxio/worker/block/TieredBlockStore.java]: `invokeAll` cancels unfinished tasks if interrupted.

Comment (ggezer) [reply core/server/worker/src/main/java/alluxio/worker/block/TieredBlockStore.java]: Caller was crash-exiting. I declared `InterruptedException` along the call chain to gracefully handle it.

File: core/server/worker/src/main/java/alluxio/worker/block/TieredBlockStore.java (hunk 5 above)
Comment (ggezer) [reply core/server/worker/src/main/java/alluxio/worker/block/TieredBlockStore.java]: `invokeAll` guarantees that all returned futures will be completed.
This is merely to satisfy method signature.

File: core/common/src/main/java/alluxio/conf/PropertyKey.java (hunk 2 above)
Comment (ggezer) [reply core/common/src/main/java/alluxio/conf/PropertyKey.java]: not really. Changed to `WORKER_EVICTOR_BLOCK_TRANSFER_THREAD_COUNT`

**Comment by ggezer** [comment]
jenkins, test this please

File: core/common/src/main/java/alluxio/conf/PropertyKey.java (hunk 1 above)
Comment (calvinjia) [reply core/common/src/main/java/alluxio/conf/PropertyKey.java]: Right, it pretty much allocates all resources on the machine to eviction if needed, that seems a bit aggressive?

File: core/server/worker/src/main/java/alluxio/worker/block/TieredBlockStore.java (hunk 7 above)
Comment (calvinjia) [reply core/server/worker/src/main/java/alluxio/worker/block/TieredBlockStore.java]: How does `InterruptedException` get handled further up the stack? If eviction is interrupted, should we just interpret that as out of space?

Comment (ggezer) [reply core/server/worker/src/main/java/alluxio/worker/block/TieredBlockStore.java]: Not throwing eviction will cause cascading eviction runs to continue until the heartbeat executor notices the interruption. And we'd need to properly store interruption status to heartbeat executor.

File: core/common/src/main/java/alluxio/conf/PropertyKey.java (hunk 1 above)
Comment (ggezer) [reply core/common/src/main/java/alluxio/conf/PropertyKey.java]: Should we leave one core out? Or half the core count is better?

Comment (calvinjia) [reply core/common/src/main/java/alluxio/conf/PropertyKey.java]: Half of the cores seems a more moderate choice, you can add some more details in the description about how this parameter can be tuned.

**Comment by ggezer** [comment]
jenkins, test this please

**Comment by ggezer** [comment]
Closing for now...

**Pull request closed without merging**